recommend auth code flow using signed requests - reloaded (#251)
camaraproject/IdentityAndConsentManagement: documentation/CAMARA-Security-Interoperability.md

The OIDC Authorization Code Flow is defined in [OpenID Connect](https://openid.net/specs/openid-connect-core-1_0.html)

### Signed Authentication Requests

> It is RECOMMENDED that signed authentication requests be used, as specified by [OIDC](https://openid.net/specs/openid-connect-core-1_0.html#JWTRequests). The same key MAY be used for signing the authentication request as is used for [client authentication](#client-authentication). 

Reviewer: ```suggestion
It is RECOMMENDED that signed authentication requests be used, as specified by [RFC9101](https://www.rfc-editor.org/rfc/rfc9101.html). The same key MAY be used for signing the authentication request as is used for [client authentication](#client-authentication). 
```

Author: A text like "as specified by [RFC9101](https://www.rfc-editor.org/rfc/rfc9101.html" would **force** readers and developers to read RFC9191. I would like to avoid that because OIDC says that:
> Note that the Request Objects defined here are compatible with those specified by [The OAuth 2.0 Authorization Framework: JWT-Secured Authorization Request (JAR)](https://openid.net/specs/openid-connect-core-1_0.html#RFC9101) [RFC9101].

How about adding a note that the security considerations of rfc9101 apply?

Author: I edited the PR so that it now states that the Security Considerations of RFC9101 apply. I think that with that text the profile does not force developers to read RFC9101 but they can trust this document that we considered all security advise.

Reviewer: You made a counter proposal, and without giving me a chance to respond, you made a new commit that changed the text again. Could we please just leave the text as it was and make the reference to RFC9101 just for the sake of reaching a final agreement? We at Telefónica have already accepted the rest of the suggestions made by both @mhfoo and you.

Author: You can also respond and have all the chances you want. Please comment make a new review change request. Noting I proposed is written in stone. 
Yes, I made a "counter proposal" and I think making proposals is the fastest way forward. I added my coment and then proposed new text. I think this procodure allows us to move forward fastest.

I think review-change requests and adaption to the PR give us a way to come to an agreement much faster than comments in issues. Please trust me that I do not want to force that new text on you or anybody. 
Please provide a new review-comment or change request. This is your chance.

Reviewer: OK, I see. It's as if I was directly committing my change rather than making it a change suggestion... but ok, I'll let this go anyway.

> It is RECOMMENDED that signed authentication requests be used, as specified by [OIDC](https://openid.net/specs/openid-connect-core-1_0.html#JWTRequests). The same key MAY be used for signing the authentication request as is used for [client authentication](#client-authentication). 

Reviewer: ```suggestion
It is RECOMMENDED that signed authentication requests be used, as specified by [OIDC](https://openid.net/specs/openid-connect-core-1_0.html#JWTRequests). The same key MAY be used for signing the authentication request as is used for [client authentication](#client-authentication). In addition, this document defines that:
  * Values for the mandatory parameters `scope`, `response_type`, `client_id` and `redirect_uri` MUST be included using the OAuth2.0 request syntax. The values for these parameters MUST match those in the signed request object.
  * For all other parameters, the authorization server MUST only use the parameters contained in the signed request object. 
```

It is RECOMMENDED that the value of the `aud` field of the signed authentication request is the URL of the [Authorization Endpoint](https://openid.net/specs/openid-connect-core-1_0.html#AuthorizationEndpoint).
The authorization server MAY accept different values of the `aud` field e.g. the `issuer` field of its [OpenID Provider Metadata](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata). 
The authorization server MUST check the value of the `aud` field and reject signed authentication requests if the value of the `aud` is not associated with the authorization server.

Note: Care must be taken in a multi-tenant environment that a signed authentication request for one tenant is not accepted at another tenant endpoint.

> Note: The [Security Considerations](https://www.rfc-editor.org/rfc/rfc9101.html#section-10) of RFC9191 apply.It is recommended that the API consumer never includes a `sub` field in the signed request object, because otherwise the signed request object might be used for client authenticaton. For security reasons the endpoint receiving the request is explicitly named in the `aud` field.

Author: Why is this suggestion so big, when the only addition is this sentence? 
> This document defines that the authorization server MUST only use the parameters included in the signed request object.

This is the only change, right?

One of the reasons for having signed request objects is that the RP can have a fixed signed request object and only vary the optional parameters like "state" and "nonce" that should be different for each request.

> this would typically be done to enable a cached, pre-signed (and possibly pre-encrypted) Request Object value to be used containing the fixed request parameters, while parameters that can vary with each request, such as state and nonce, are passed as OAuth 2.0 parameters.

What is the threat this security measure is mitigating?
What would an attacker gain by adding prompt=consent?

I am all in favor for integrity protected requests.

> Note: The [Security Considerations](https://www.rfc-editor.org/rfc/rfc9101.html#section-10) of RFC9191 apply.It is recommended that the API consumer never includes a `sub` field in the signed request object, because otherwise the signed request object might be used for client authenticaton. For security reasons the endpoint receiving the request is explicitly named in the `aud` field.

Reviewer: > This is the only change, right?

Yes it is, and a couple of typos in line 104.

> What is the threat this security measure is mitigating?
- Parameters outside the signed object can be manipulated by attackers, changing the original intent of the request.
- Conflicting or malicious values in the unsigned parameters can cause the authorisation server to behave incorrectly.
- Accepting external parameters opens the door to injection or manipulation attacks.

This measure is aligned with [RFC9101 (JAR)](https://www.rfc-editor.org/rfc/rfc9101.html#section-5) also considered in specifications like [FAPI Security Profile 1.0 - Part 2: Advanced](https://openid.net/specs/openid-financial-api-part-2-1_0.html#authorization-server)

> What would an attacker gain by adding prompt=consent?

It was just an example, different from the happy paths examples provided [here](https://github.com/camaraproject/IdentityAndConsentManagement/pull/226#issuecomment-2602485710), to point out that "prompt=consent" would affect the way the authorisation server handles the request with respect to its original intent.

> Note: The [Security Considerations](https://www.rfc-editor.org/rfc/rfc9101.html#section-10) of RFC9191 apply.It is recommended that the API consumer never includes a `sub` field in the signed request object, because otherwise the signed request object might be used for client authenticaton. For security reasons the endpoint receiving the request is explicitly named in the `aud` field.

Author: I just wanted to make clear, that cached, pre-signed request objects are then not allowed/possible any more.

> this would typically be done to enable a cached, pre-signed (and possibly pre-encrypted) Request Object value to be used containing the fixed request parameters, while parameters that can vary with each request, such as state and nonce, are passed as OAuth 2.0 parameters.

And required parameters are still required even though they are not used, right?
A request without scope, response_type, client_id and redirect_uri is invalid.
Should we RECOMMEND that the values are the same?

This example would be invalid because the required parameters are missing, right?

```
GET /authorize?
request=base64urlencode(signed({
"response_type": "code",
"scope": "openid%20profile%20email",
"client_id": "s6BhdRkqt3",
"redirect_uri": "https%3A%2F%2Fclient.example.org%2Fcb"
})
```

But this example would be valid, if the request object values are valid:
```
GET /authorize?response_type=abc&scope=def&client_id=ghi&redirect_uri=https://foo.org/&
request=base64urlencode(signed({
"response_type": "code",
"scope": "openid%20profile%20email",
"client_id": "s6BhdRkqt3",
"redirect_uri": "https%3A%2F%2Fclient.example.org%2Fcb"
})
```

> Note: The [Security Considerations](https://www.rfc-editor.org/rfc/rfc9101.html#section-10) of RFC9191 apply.It is recommended that the API consumer never includes a `sub` field in the signed request object, because otherwise the signed request object might be used for client authenticaton. For security reasons the endpoint receiving the request is explicitly named in the `aud` field.

Author: Actually, I do not find the part of the spec that says the required parameters cannot be in the request object alone. 

If nobody finds this, then the above comment would be reduced to this:
> I just wanted to make clear, that cached, pre-signed request objects are then not allowed/possible any more.

> this would typically be done to enable a cached, pre-signed (and possibly pre-encrypted) Request Object value to be used containing the fixed request parameters, while parameters that can vary with each request, such as state and nonce, are passed as OAuth 2.0 parameters.

If we agree that the new sentence prevents cached, pre-signed request objects then the new sentence is ok and I would approve this PR.

> Note: The [Security Considerations](https://www.rfc-editor.org/rfc/rfc9101.html#section-10) of RFC9191 apply.It is recommended that the API consumer never includes a `sub` field in the signed request object, because otherwise the signed request object might be used for client authenticaton. For security reasons the endpoint receiving the request is explicitly named in the `aud` field.

Reviewer: From the Telefónica side, our proposed text would be fine as is, but considering that other API providers may use existing commercial frameworks to implement OIDC, it may result in an invalid request if any of the mandatory parameters (scope, response_type, client_id and redirect_uri) are not present outside the signed request object. So we could add an additional clarification, similar to what is done in [Mobile Connect](https://www.gsma.com/solutions-and-impact/technologies/mobile-identity/wp-content/uploads/2022/11/IDY.02-v2.0.pdf) for example.

> Values for the scope, response_type and client_id parameters MUST be
included using the OAuth2.0 request syntax. The values for these parameters
MUST match those in the Request Object; i.e.: these parameters MUST exist both
in the OIDC Authorization request and signed request object. Comparing both the
values will yield more security.

i.e. Mandatory OIDC parameters must be provided and their values must match those in the request object. Mobile Connect does not mention `redirect_uri` though.

[OIDC](https://openid.net/specs/openid-connect-core-1_0.html#RequestObject) also states:

> So that the request is a valid OAuth 2.0 Authorization Request, values for the response_type and client_id parameters MUST be included using the OAuth 2.0 request syntax, since they are REQUIRED by OAuth 2.0. The values for these parameters MUST match those in the Request Object, if present.
>
>Even if a scope parameter is present in the Request Object value, a scope parameter MUST always be passed using the OAuth 2.0 request syntax containing the openid scope value to indicate to the underlying OAuth 2.0 logic that this is an OpenID Connect request.

> Note: The [Security Considerations](https://www.rfc-editor.org/rfc/rfc9101.html#section-10) of RFC9191 apply.It is recommended that the API consumer never includes a `sub` field in the signed request object, because otherwise the signed request object might be used for client authenticaton. For security reasons the endpoint receiving the request is explicitly named in the `aud` field.

Author: Would you please create suggestion? (And remove the current big one).
Then I could click the "commit suggestion" button, and I do not get complaints that I change the PR to quickly.

> Note: The [Security Considerations](https://www.rfc-editor.org/rfc/rfc9101.html#section-10) of RFC9191 apply.It is recommended that the API consumer never includes a `sub` field in the signed request object, because otherwise the signed request object might be used for client authenticaton. For security reasons the endpoint receiving the request is explicitly named in the `aud` field.

Author: Or we leave the PR as it is recommending signed request object and don't force anything on implementors that their chosen implementation/framework might not support

> Note: The [Security Considerations](https://www.rfc-editor.org/rfc/rfc9101.html#section-10) of RFC9191 apply.It is recommended that the API consumer never includes a `sub` field in the signed request object, because otherwise the signed request object might be used for client authenticaton. For security reasons the endpoint receiving the request is explicitly named in the `aud` field.

Reviewer: ```suggestion
Note: The [Security Considerations](https://www.rfc-editor.org/rfc/rfc9101.html#section-10) of RFC9191 apply. It is recommended that the API consumer never includes a `sub` field in the signed request object, because otherwise the signed request object might be used for client authentication. For security reasons the endpoint receiving the request is explicitly named in the `aud` field.
```

> Note: The [Security Considerations](https://www.rfc-editor.org/rfc/rfc9101.html#section-10) of RFC9191 apply.It is recommended that the API consumer never includes a `sub` field in the signed request object, because otherwise the signed request object might be used for client authenticaton. For security reasons the endpoint receiving the request is explicitly named in the `aud` field.

Reviewer: > Would you please create a suggestion? (And remove the current big one). Then I could click the "commit suggestion" button, and I do not get complaints that I change the PR to quickly.

Done. https://github.com/camaraproject/IdentityAndConsentManagement/pull/251#discussion_r1926717968

> Note: The [Security Considerations](https://www.rfc-editor.org/rfc/rfc9101.html#section-10) of RFC9191 apply.It is recommended that the API consumer never includes a `sub` field in the signed request object, because otherwise the signed request object might be used for client authenticaton. For security reasons the endpoint receiving the request is explicitly named in the `aud` field.

Reviewer: > Or we leave the PR as it is recommending signed request object and don't force anything on implementors that their chosen implementation/framework might not support

For us, the statement about not using parameters not included in the signed request object is important. Simply adding "_This document defines that the authorization server MUST only use the parameters included in the signed request object._" would also work for us. As suggested [above](https://github.com/camaraproject/IdentityAndConsentManagement/pull/251#discussion_r1923866083).
